tests/helpers.py

```diff
@@ -160,3 +166,174 @@ def __init__(self, **kwargs):
+    validation_fail: list[tuple[dict[str, Any], type[Exception]]] = []
+
+    # This holds the parameters for the deserialization error tests
+    deserialization_fail: list[tuple[bytes, type[Exception]]] = []
+
+    for data_or_exc, expected_bytes_or_exc in test_data:
+        if isinstance(data_or_exc, tuple) and isinstance(expected_bytes_or_exc, bytes):
+            kwargs = dict(zip([f[0] for f in fields], data_or_exc))
+            parameters.append((kwargs, expected_bytes_or_exc))
+        elif isinstance(data_or_exc, type) and isinstance(expected_bytes_or_exc, bytes):
+            deserialization_fail.append((expected_bytes_or_exc, data_or_exc))
+        else:
+            data = cast(Tuple[Any, ...], data_or_exc)
+            exception = cast(type[Exception], expected_bytes_or_exc)
+            kwargs = dict(zip([f[0] for f in fields], data))
+            validation_fail.append((kwargs, exception))
+
+    def generate_name(param: dict[str, Any] | bytes, i: int) -> str:
+        """Generate a name for the test case."""
+        length = 30
+        result = f"{i:02d}] : "  # the first [ is added by pytest
+        if isinstance(param, bytes):
+            try:
+                result += str(param[:length], "utf-8") + "..." if len(param) > (length + 3) else param.decode("utf-8")
+            except UnicodeDecodeError:
+                result += repr(param[:length]) + "..." if len(param) > (length + 3) else repr(param)
+        else:
+            param = cast(Dict[str, Any], param)
+            begin = ", ".join(f"{k}={v}" for k, v in param.items())
+            result += begin[:length] + "..." if len(begin) > (length + 3) else begin
+        result = result.replace("\n", "\\n").replace("\r", "\\r")
+        result += f" [{cls.__name__}"  # the other [ is added by pytest
+        return result
+
+    class TestClass:
+        """Test class for the generated tests."""
+
+        @pytest.mark.parametrize(
+            ("kwargs", "expected_bytes"),
+            parameters,
+            ids=tuple(generate_name(kwargs, i) for i, (kwargs, _) in enumerate(parameters)),
+        )
+        def test_serialization(self, kwargs: dict[str, Any], expected_bytes: bytes):
+            """Test serialization of the object."""
+            obj = cls(**kwargs)
+            serialized_bytes = obj.serialize()
+            assert serialized_bytes == expected_bytes, f"{serialized_bytes} != {expected_bytes}"
+
+        @pytest.mark.parametrize(
+            ("kwargs", "expected_bytes"),
+            parameters,
+            ids=tuple(generate_name(kwargs, i) for i, (kwargs, _) in enumerate(parameters)),
+        )
+        def test_deserialization(self, kwargs: dict[str, Any], expected_bytes: bytes):
+            """Test deserialization of the object."""
+            buf = Buffer(expected_bytes)
+            obj = cls.deserialize(buf)
+            assert cls(**kwargs) == obj, f"{cls.__name__}({kwargs}) != {obj}"
```

Comment:
I wonder if it'd be better to go over each kwarg individually instead for better error messages. Also, it'd help in case some class overrides `__eq__` and drops some check, because of some custom behavior with types of that kind, which however doesn't apply to these deserialization tests.

Reply:
I could try to find the mismatch in the kwargs, but the classes don't **have to** have their kwargs as attributes, so I think it should only be an additionnal information (that's what I like about this system, we can have really nice tests and messages without making much effort).

I think using `__eq__` is good because this method should really return `True` for objects that should be considered the same, so if the client recieves an object and sends it back, even if for example the order of the `CompoundTag` has changed, they should be considered equal on the server because they are.

Reply:
Okay, but mention that `__eq__` is being used rather than comparing the exact params in the docstring of the `gen_serializable_test` func, as that may have implications on how it's used.

Reply:
Ok I'll amend that in